File: IntrinsicsCSharp/SumProduct.cs
using System;
using System.Linq;
using System.Runtime.Intrinsics;
using System.Runtime.Intrinsics.X86;
using System.Threading.Tasks;

using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

namespace NetCoreIntrinsics
{
    public class SumProduct
    {
        double[] Data1, Data2;
        int n = 1000000;

        [GlobalSetup]
        public void InitialiseData()
        {
            Data1 = new double[n];
            Data2 = new double[n];

            for (int i = 0; i < n; i++)
            {
                Data1[i] = 10.0 * Math.Sin(i) + Math.Log(i + 10.0);
                Data2[i] = 15.0 * (0.5 + Math.Cos(0.25 * i));
            }
        }

        [Benchmark]
        public double WithLinq()
        {
            return Data1.Zip(Data2, (x1, x2) => x1 * x2).Sum();
        }

        [Benchmark]
        public double AccumulateFor()
        {
            double acc = 0.0;

            for (int i = 0; i < n; i++)
            {
                acc += Data1[i] * Data2[i];
            }

            return acc;
        }

        [Benchmark]
        public double ParallelFor()
        {
            var degrees = 4;

            var sums = new double[degrees];

            Action<int> SumChunk = chunkIndex =>
            {
                double sum = 0.0;

                for (int i = chunkIndex * n / degrees; i < (chunkIndex + 1) * n / degrees; i++)
                {
                    sum += Data1[i] * Data2[i];
                }

                sums[chunkIndex] = sum;
            };

            Parallel.For(0, degrees, SumChunk);

            return sums.Sum();
        }

        [Benchmark]
        public unsafe double WithSse2()
        {
            double result;

            fixed (double* d1 = Data1, d2 = Data2)
            {
                var mul = Vector128<double>.Zero;
                var acc = Vector128<double>.Zero;

                for (int i = 0; i < n; i += 2)
                {
                    mul = Sse2.Multiply(Sse2.LoadVector128(d1 + i), Sse2.LoadVector128(d2 + i));
                    acc = Sse2.Add(acc, mul);
                }

                result = acc.GetElement(0) + acc.GetElement(1);
            }

            return result;
        }

        [Benchmark]
        public unsafe double WithAvx()
        {
            double result;

            fixed (double* d1 = Data1, d2 = Data2)
            {
                var mul = Vector256<double>.Zero;
                var acc = Vector256<double>.Zero;

                for (int i = 0; i < n; i += 4)
                {
                    mul = Avx.Multiply(Avx.LoadVector256(d1 + i), Avx.LoadVector256(d2 + i));
                    acc = Avx.Add(acc, mul);
                }

                result = acc.GetElement(0) + acc.GetElement(1) + acc.GetElement(2) + acc.GetElement(3);
            }

            return result;
        }

        [Benchmark]
        public unsafe double ParallelAvx()
        {
            var degrees = 4;

            var sums = new double[degrees];

            Action<int> SumChunk = chunkIndex =>
            {
                fixed (double* d1 = Data1, d2 = Data2)
                {
                    var mul = Vector256<double>.Zero;
                    var acc = Vector256<double>.Zero;

                    for (int i = chunkIndex * n / degrees; i < (chunkIndex + 1) * n / degrees; i += 4)
                    {
                        mul = Avx.Multiply(Avx.LoadVector256(d1 + i), Avx.LoadVector256(d2 + i));
                        acc = Avx.Add(acc, mul);
                    }

                    sums[chunkIndex] = acc.GetElement(0) + acc.GetElement(1) + acc.GetElement(2) + acc.GetElement(3);
                }
            };

            Parallel.For(0, degrees, SumChunk);

            return sums.Sum();
        }

        static void Main(string[] args)
        {
            var sumProduct = new SumProduct();
            sumProduct.InitialiseData();

            Console.WriteLine($"WithLinq: {sumProduct.WithLinq():F4}");
            Console.WriteLine($"AccumulateFor: {sumProduct.AccumulateFor():F4}");
            Console.WriteLine($"ParallelFor: {sumProduct.ParallelFor():F4}");
            Console.WriteLine($"WithSse2: {sumProduct.WithSse2():F4}");
            Console.WriteLine($"WithAvx: {sumProduct.WithAvx():F4}");
            Console.WriteLine($"ParallelAvx: {sumProduct.ParallelAvx():F4}");
            Console.ReadLine();

            BenchmarkRunner.Run<SumProduct>();
        }
    }
}

File: IntrinsicsFSharp/SumProduct.fs
module TryIntrinsics.SumProduct

open System.Threading.Tasks

let map2 (data1: float[]) (data2: float[]) =
    Array.map2 (fun x y -> x * y) data1 data2
    |> Array.sum

let accumulateIter (data1: float[]) (data2: float[]) =
    let mutable acc = 0.0

    data1
    |> Array.iteri (fun n x ->
        acc <- acc + x * data2.[n])
    acc

let accumulateFor (data1: float[]) (data2: float[]) =
    let mutable acc = 0.0

    for i in 0 .. (data1.Length - 1) do
        acc <- acc + data1.[i] * data2.[i]
    acc

let naiveParallelFor (data1: float[]) (data2: float[]) =
    let sums = Array.zeroCreate data1.Length

    Parallel.For(0, data1.Length, fun i ->
        sums.[i] <- data1.[i] * data2.[i]) |> ignore

    Array.sum sums

let middlingParallelFor (data1: float[]) (data2: float[]) =
    let degrees = 4
    let sums = Array.zeroCreate degrees

    Parallel.For(0, degrees, fun p ->
        let mutable acc = 0.0

        for i in p .. degrees .. (data1.Length - 1)  do
            acc <- acc + data1.[i] * data2.[i]
        
        sums.[p] <- acc) |> ignore

    let mutable acc = 0.0
    for i in 0 .. (sums.Length - 1) do
        acc <- acc + sums.[i]
    acc

let smartParallelFor (data1: float[]) (data2: float[]) =
    let degrees = 4
    let sums = Array.zeroCreate degrees

    Parallel.For(0, degrees, fun p ->
        let mutable acc = 0.0
        let startIndex = data1.Length * p / degrees
        let endIndex =
            if p = degrees - 1 then
                data1.Length - 1
            else
                data1.Length * (p + 1) / degrees - 1

        for i in startIndex .. endIndex do
            acc <- acc + data1.[i] * data2.[i]
        
        sums.[p] <- acc) |> ignore

    let mutable acc = 0.0
    for i in 0 .. (sums.Length - 1) do
        acc <- acc + sums.[i]
    acc

open System.Runtime.Intrinsics
open System.Runtime.Intrinsics.X86

let withSse2 (data1: float[]) (data2: float[]) =

    if not Sse2.IsSupported then failwith "SSE2 not available"

    use d1 = fixed data1
    use d2 = fixed data2

    let mutable accV = Vector128<float>.Zero // holds two double-precision floats
    let mutable mulV = Vector128<float>.Zero
    
    let lastIndex = data1.Length - (data1.Length / 2) - 2

    for i in 0 .. 2 .. lastIndex do
        mulV <- Sse2.Multiply(Sse2.LoadVector128(NativeInterop.NativePtr.add d1 i), Sse2.LoadVector128(NativeInterop.NativePtr.add d2 i))
        accV <- Sse2.Add(accV, mulV)

    let mutable acc = accV.GetElement(0) + accV.GetElement(1)

    // any remaining elements
    for i in lastIndex + 1 .. data1.Length - 1 do
        acc <- acc + data1.[i] * data2.[i]

    acc

let withSse2B (data1: float[]) (data2: float[]) =

    if not Sse2.IsSupported then failwith "SSE2 not available"

    use d1 = fixed data1
    use d2 = fixed data2

    let mutable accV = Vector128<float>.Zero // holds two double-precision floats
    let mutable mulV = Vector128<float>.Zero
    
    let lastIndex = data1.Length - (data1.Length / 2) - 2

    for i in 0 .. 2 .. lastIndex do
        mulV <- Sse2.Multiply(Sse2.LoadVector128(NativeInterop.NativePtr.add d1 i), Sse2.LoadVector128(NativeInterop.NativePtr.add d2 i))
        accV <- Sse2.Add(accV, mulV)

    let mutable acc = accV.GetElement(0) + accV.GetElement(1)

    // any remaining elements
    for i in lastIndex + 1 .. data1.Length - 1 do
        acc <- acc + data1.[i] * data2.[i]

    acc

let withSse3 (data1: float[]) (data2: float[]) =

    if not Sse3.IsSupported then failwith "SSE3 not available"

    use d1 = fixed data1
    use d2 = fixed data2

    let mutable accV = Vector128<float>.Zero // holds two double-precision floats
    let mutable mulV = Vector128<float>.Zero
    
    let lastIndex = data1.Length - (data1.Length / 2) - 2

    for i in 0 .. 2 .. lastIndex do
        mulV <- Sse3.Multiply(Sse3.LoadVector128(NativeInterop.NativePtr.add d1 i), Sse3.LoadVector128(NativeInterop.NativePtr.add d2 i))
        accV <- Sse3.Add(accV, mulV)

    let mutable acc = accV.GetElement(0) + accV.GetElement(1)

    // any remaining elements
    for i in lastIndex + 1 .. data1.Length - 1 do
        acc <- acc + data1.[i] * data2.[i]

    acc

let withAvx (data1: float[]) (data2: float[]) =

    if not Avx.IsSupported then failwith "AVX not available"

    use d1 = fixed data1
    use d2 = fixed data2

    let mutable accV = Vector256<float>.Zero // holds four double-precision floats
    let mutable mulV = Vector256<float>.Zero
    
    let lastIndex = data1.Length - (data1.Length / 4) - 4

    for i in 0 .. 4 .. lastIndex do
        mulV <- Avx.Multiply(Avx.LoadVector256(NativeInterop.NativePtr.add d1 i), Avx.LoadVector256(NativeInterop.NativePtr.add d2 i))
        accV <- Avx.Add(accV, mulV)

    let mutable acc = accV.GetElement(0) + accV.GetElement(1) + accV.GetElement(2) + accV.GetElement(3)

    // any remaining elements
    for i in lastIndex + 1 .. data1.Length - 1 do
        acc <- acc + data1.[i] * data2.[i]

    acc

let withAvx2 (data1: float[]) (data2: float[]) =

    if not Avx2.IsSupported then failwith "AVX2 not available"

    use d1 = fixed data1
    use d2 = fixed data2

    let mutable accV = Vector256<float>.Zero // holds four double-precision floats
    let mutable mulV = Vector256<float>.Zero
    
    let lastIndex = data1.Length - (data1.Length / 4) - 4

    for i in 0 .. 4 .. lastIndex do
        mulV <- Avx2.Multiply(Avx2.LoadVector256(NativeInterop.NativePtr.add d1 i), Avx2.LoadVector256(NativeInterop.NativePtr.add d2 i))
        accV <- Avx2.Add(accV, mulV)

    let mutable acc = accV.GetElement(0) + accV.GetElement(1) + accV.GetElement(2) + accV.GetElement(3)

    // any remaining elements
    for i in lastIndex + 1 .. data1.Length - 1 do
        acc <- acc + data1.[i] * data2.[i]

    acc

let withFma (data1: float[]) (data2: float[]) =

    if not Fma.IsSupported then failwith "FMA not available"

    use d1 = fixed data1
    use d2 = fixed data2

    let mutable accV = Vector256<float>.Zero // holds four double-precision floats
    let mutable mulV = Vector256<float>.Zero
    
    let lastIndex = data1.Length - (data1.Length / 4) - 4

    for i in 0 .. 4 .. lastIndex do
        accV <- Fma.MultiplyAdd(accV, Fma.LoadVector256(NativeInterop.NativePtr.add d1 i), Fma.LoadVector256(NativeInterop.NativePtr.add d2 i))

    let mutable acc = accV.GetElement(0) + accV.GetElement(1) + accV.GetElement(2) + accV.GetElement(3)

    // any remaining elements
    for i in lastIndex + 1 .. data1.Length - 1 do
        acc <- acc + data1.[i] * data2.[i]

    acc